### PR TITLE
Unify derivative function signatures

### PR DESCRIPTION
**Pull request by dschwoerer**
Most accept now CELL_LOC, DIFF_METHOD and REGION.
Some are still ignoring these

Possible solution for #811

### CODE REVIEW

**Comment by dschwoerer** [comment]
It is still possible to change everything from hand - I just thought I add the script I used to generate that.
If it is preferred, I could have a go with jinja, to use that to generate it, or just remove the `makefile` and the `.in` file, then it looks as if I had changed everything from hand 👍

File: include/derivs.hxx.in [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+inline const Field3D DDZ(const Field3D &f, DIFF_METHOD method, CELL_LOC outloc, bool inc_xbndry) {
+  return DDZ(f, outloc, method, inc_xbndry ? RGN_NOX : RGN_NOBNDRY);
+}
+
+inline const Field3D DDZ(const Field3D &f, DIFF_METHOD method, bool inc_xbndry) {
+  return DDZ(f, CELL_DEFAULT, method, inc_xbndry ? RGN_NOX : RGN_NOBNDRY);
+}
+
+inline const Field3D DDZ(const Field3D &f, bool inc_xbndry) {
+  return DDZ(f, CELL_DEFAULT, DIFF_DEFAULT, inc_xbndry ? RGN_NOX : RGN_NOBNDRY);
```
Comment (d7919) [inline include/derivs.hxx.in]: Shouldn't this be RGN_NOY if we ask to include the x boundary?

**Comment by d7919** [comment]
There's nearly 5,000 additions to the header (which is currently just ~360 lines long) which is perhaps the surprising part given that currently the longest file is less than 3.5k lines. I guess this is related to the repeated documentation issue that you have linked earlier.

I think if we're going to go with code generation then it's good to use the same approach across the library.

**Comment by dschwoerer** [comment]
I have merged this into #641 to expose the derivatives

**Comment by dschwoerer** [comment]
I noticed that the `test-code-style` was disabled - probably since #637, the move to the new testing framework

**Comment by d7919** [comment]
I don't think it is disabled -- travis reports that the test is run https://travis-ci.org/boutproject/BOUT-dev/jobs/338634976#L1865

**Comment by dschwoerer** [comment]
Sorry for not being clearer.
It is not disabled, but It is useless, as it only checks `$BOUT_TOP/tests` rather then `$BOUT_TOP`
[This](https://github.com/boutproject/BOUT-dev/pull/815/commits/603a17af8980ec4a476f6524f161526ca1054dc0#diff-3a4692034e53243fc83eec06f6b549fbL3) should fix it

**Comment by dschwoerer** [comment]
#848 needs to applied here, if it is accepted

**Comment by dschwoerer** [comment]
This is broken because the code style test is disabled in next, and therefore nobody noticed this ...
@d7919 - do you mind fixing this?

**Comment by d7919** [comment]
Thanks for catching this, fixed in #910

**Comment by ZedThree** [comment]
The main thing stopping this going in at the minute is the size -- it's hard to understand.

One reason it's so big is that this generates all possible permutations (6) of the last three arguments:

```cpp
DDX(const Field3D &f, CELL_LOC outloc, DIFF_METHOD method, REGION region)
```

I'm not sure there's any need for that. We currently allow different orders of `method` and `outloc`, but I propose we deprecate those, and only use the order shown above or with `region` and `method` reversed. That would significantly reduce the complexity of this PR.

We don't use any of the "reordered" versions in the tests or examples, so I suspect they are only there for historical reasons.

**Comment by dschwoerer** [comment]
So should I just keep
```
/* Comment */
DDX(const Field3D &f, CELL_LOC outloc, REGION region, DIFF_METHOD method)
/* no comment? */
DEPRECATED(DDX(const Field3D &f, REGION region, CELL_LOC outloc, DIFF_METHOD method))
```
Especially just having the comment just once will make the file significantly smaller ...

**Comment by ZedThree** [comment]
Yes, that's good. Keep the default arguments for `outloc, region, method` too, in case that wasn't obvious. Switch to C++ style comments too, so end result is:

```cpp
/// Calculate first partial derivative in X
///
///   $\partial / \partial x$
///
/// @param[in] f       The field to be differentiated
/// @param[in] outloc  The cell location where the result is desired. If
///                    staggered grids is not enabled then this has no effect
///                    If not given, defaults to CELL_DEFAULT
/// @param[in] method  Differencing method to use. This overrides the default
///                    If not given, defaults to DIFF_DEFAULT
/// @param[in] region  What region is expected to be calculated
///                    If not given, defaults to RGN_NOBNDRY
const Field3D DDX(const Field3D &f, CELL_LOC outloc = CELL_DEFAULT,
                  DIFF_METHOD method = DIFF_DEFAULT, REGION region = RGN_NOBNDRY);
```

You might find then that we don't need the code generation, as you'll be cutting down on code.

**Comment by dschwoerer** [comment]
We didn't need it now - it just makes such changes easier ...
We did not have it before, and only few functions had comments.
I mentioned already in the beginning - if you do not like code generation as such - I can always just not check in the generating code ... but it makes my life much easier. And I am sad that I have to change many function signatures by hand in `derivs.cxx` - Maybe I should also use code generation there ...

**Comment by dschwoerer** [comment]
I disagree - I think `derivs.hxx` should be auto generated.
 * It has been ported to jinja2 - so that it is easily mentainable, after you requested it
 * The file is highly repetitive - this suitable for autogeneration
 * The only exceptions are the header with includes and some deprecated methods at the end.
 * We will have to mass edit this file - if we remove all the deprecated methods
 * By using the template we basically ensure that all functions stay aligned - which is IMHO essential to keep a clean and unified interface for BOUT++ - making it as surprise free as possible.

**Comment by ZedThree** [comment]
It's repetitive, but not complicated. It also doesn't change very often, as it's the public API. Removing all the deprecated functions is a trivial sed one-liner, for example. Like I said, it's a trade off between making our lives easier and having to maintain more complicated tools. In this particular case, the code is so simple, that the complicated tools aren't necessary to keep around.

It's worth trying to keep things simple where possible, so let's see if we can simplify things here enough that we don't need the code generation. 

---

Changing the definition of `DEPRECATED` to:

```cpp
#define DEPRECATED(func) __attribute__ ((deprecated)) func
```

lets us use it on function definitions. I've just tested this on a simple test case with gcc 4.8.3 (default `-std=c++98` ), 7.1.0, clang 3.8.0, 4.0.0 and intel 17.0.1, so this should be fine. With gcc 7.3.1, it doesn't cause problems in BOUT++.

Doing this, we can then just use

```cpp
DEPRECATED(inline const Field3D DDX(const Field3D &f, DIFF_METHOD method,
                                    CELL_LOC outloc = CELL_DEFAULT,
                                    REGION region = RGN_NOBNDRY)) {
  return DDX(f, outloc, method, region);
}
```

instead of having the declaration followed by the definition.

Also note that in it's current form we're getting deprecated warnings for [existing uses of some derivatives][1] -- note that this is the "correct" way round, so we don't want to deprecate this one:

```cpp
DEPRECATED(const Field3D DDY(const Field3D &f, CELL_LOC outloc, DIFF_METHOD method,
                             REGION region));
```

We should be able to cut down the number of overloads to just two per functions: the one we want, with default values for `outloc, method, region`, and the deprecated one with `method, outloc`. This basically cuts the length of `derivs.hxx` in half.

---

It's still not obvious to me which way round we want the `method` and `region` arguments. I've been playing about with it a bit this morning, and either way round has some issues. It turns out there's a couple of places where we haven't been consistent.

[1]: https://travis-ci.org/boutproject/BOUT-dev/jobs/355956220#L1213

**Comment by dschwoerer** [comment]
I misread you the last time, and thought you wanted the `DDX(const Field3D &f, CELL_LOC outloc, REGION region, DIFF_METHOD method)` version - that is why I changed `derivs.cxx`.

As changing that back would take more effort - could you decide on what you want?
Depending on that I will undo that commit, or keep it, and then change the DEPRECATED macro.
But if I do it now I have to do everything again, or end up in the hell of merge-conflicts ...

PS: With the ordering as above, we need to keep two deprecated versions ...

----------------------------

Concerning the keep of the auto-generated files: I guess much can be done with tools like sed and the like. However if I look at more complicated files, they are not mantained by tools as they have diverged.
I would struggle to remove all the deprecated functions with sed, and do it some other way - not sure that is why I prefer then jinja approach - but I guess it would be for most people easier to remove the lines from `derivs.hxx.in.function` rather than write the sed one liner ...

**Comment by dschwoerer** [comment]
I hope I have now addressed all changes:
 * Only 2 function signatures, of which one is deprecated
 * Autogeneration is kept for now, but can be removed if it turns out to be easier to modify the generated file. Thus I hope there should be no disadvantage from a maintenance point of view.
* `derivs.hxx` is down to 1060 lines (972 sloc)

**Comment by dschwoerer** [comment]
@ZedThree thanks for reviewing :+1: 

It is annoying that resolving a merge conflict dismisses the review :/

File: tests/integrated/test-code-style/runtest [hunk 2]
```diff
@@ -15,6 +15,38 @@ grep [\(,][[:space:]]*const[[:space:]][[:alnum:]]*[[:space:]][[:alnum:]]*[[:spac
     echo "Declaration of functions shouldn't declare pass-by-value as const"
 
 
+GIT=git
+if $GIT status &>/dev/null
+then
+    # Find autogenerated files
+    autogens=$(for f in $(find $paths |grep '\.in\.[a-z0-9A-Z]*$')
```
Comment (ZedThree) [inline tests/integrated/test-code-style/runtest]: This doesn't appear to work -- there's no `BOUT_TOP` in here, and you'd need to loop over `$paths`. The test passes though.

**Comment by ZedThree** [comment]
Don't worry about merge conflicts -- I can resolve those and push directly to next :) I did spot a bug in `test-code-style` though

File: tests/integrated/test-code-style/runtest (hunk 2 above)
Comment (dschwoerer) [reply tests/integrated/test-code-style/runtest]: Should be fixed now

**Pull request merged**